### PR TITLE
`sr25519` signature verification

### PR DESCRIPTION
**Pull request by kziemianek**
Add ability to call contract's `sr25519_verify` function.

`sr25519_verify` is still unstable so it's required to run node with `type UnsafeUnstableInterface = ConstBool<true>;` in pallet-contracts config.

This is related to https://github.com/paritytech/substrate/issues/13703.

### CODE REVIEW

File: crates/env/src/api.rs [hunk 1]
```diff
@@ -550,6 +550,40 @@ pub fn ecdsa_to_eth_address(pubkey: &[u8; 33], output: &mut [u8; 20]) -> Result<
     })
 }
 
+/// Verifies a sr25519 signature
```
Comment (cmichi) [inline crates/env/src/api.rs:553]:
```suggestion
/// Verifies a sr25519 signature.
```

File: crates/env/src/backend.rs [hunk 2]
```diff
@@ -290,6 +290,13 @@ pub trait EnvBackend {
         output: &mut [u8; 20],
     ) -> Result<()>;
 
+    fn sr25519_verify(
```
Comment (cmichi) [inline crates/env/src/backend.rs:293]: Please add a comment in the style of the other functions above.

File: crates/env/src/engine/off_chain/impls.rs [hunk 3]
```diff
@@ -333,6 +339,21 @@ impl EnvBackend for EnvInstance {
         Ok(())
     }
 
+    fn sr25519_verify(
+        &mut self,
+        signature: &[u8; 64],
+        message: &[u8],
+        pub_key: &[u8; 32],
+    ) -> Result<()> {
+        let context = signing_context(b"substrate");
```
Comment (kziemianek) [inline crates/env/src/engine/off_chain/impls.rs:348]: For further use, i think there should be an option to verify signature from any context. Right now there is such limitation.

**Comment by goastler** [comment]
Hey, was just testing out your changes here which solve my sr25519 signing issue in the ink smart contracts - many thanks for your changes. Just a quick fyi that if the public key is invalid the error is not caught and causes a panic (in the smart contract call, not sure where the panic originates). 

I found this from one of my unit tests (https://github.com/prosopo/protocol/blob/81c90b2edde0b71072b1f56b5b3af65a5589e9dc/contracts/lib.rs#L2292) which deliberately uses an incorrect public key. Produces the following error:
```

---- prosopo::tests::test_verify_sr25519_invalid_public_key stdout ----
data_hash: [70, 251, 116, 8, 212, 242, 133, 34, 143, 74, 245, 22, 234, 37, 133, 27]
data_hex: "46fb7408d4f285228f4af516ea25851b"
payload: <Bytes>0x46fb7408d4f285228f4af516ea25851b</Bytes>
payload_hex: 3c42797465733e307834366662373430386434663238353232386634616635313665613235383531623c2f42797465733e
signature: 0a7da2b631704cdcfe93c740e41217b9ac667a0c8755d8da1a8232db527f487c87e780d2edc1896aeb6b1bef0bc7c38d9df2135b633eab8bfb1777e82fad3a8f
caller AccountId([213, 53, 147, 199, 21, 253, 211, 28, 97, 20, 26, 189, 4, 169, 159, 214, 130, 44, 133, 88, 133, 76, 205, 227, 154, 86, 132, 231, 165, 109, 162, 125])
sig [10, 125, 162, 182, 49, 112, 76, 220, 254, 147, 199, 64, 228, 18, 23, 185, 172, 102, 122, 12, 135, 85, 216, 218, 26, 130, 50, 219, 82, 127, 72, 124, 135, 231, 128, 210, 237, 193, 137, 106, 235, 107, 27, 239, 11, 199, 195, 141, 157, 242, 19, 91, 99, 62, 171, 139, 251, 23, 119, 232, 47, 173, 58, 143]
payload [60, 66, 121, 116, 101, 115, 62, 48, 120, 52, 54, 102, 98, 55, 52, 48, 56, 100, 52, 102, 50, 56, 53, 50, 50, 56, 102, 52, 97, 102, 53, 49, 54, 101, 97, 50, 53, 56, 53, 49, 98, 60, 47, 66, 121, 116, 101, 115, 62]
thread 'prosopo::tests::test_verify_sr25519_invalid_public_key' panicked at 'called `Result::unwrap()` on an `Err` value: PointDecompressionError', /home/geopro/.cargo/git/checkouts/ink-b050680cd543eb5a/9ca19c4/crates/env/src/engine/off_chain/impls.rs:350:68
```

Happily, incorrect signatures / payload data result in verify returning false. The valid verification test also passes.

**Comment by kziemianek** [comment]
> any update on this? @cmichi @kziemianek :)

I'm a bit busy lately. I doubt I'll have a time to look at it in the near future.

**Comment by forgetso** [comment]
@cmichi Would it be possible to merge this given that the review suggestions have been completed? Obviously, after master is merged in. Thx

**Comment by SkymanOne** [comment]
Can you please address the unhandled error resulting in the call panic which was addressed by @goastler ?

**Comment by SkymanOne** [comment]
Hey @kziemianek did you have a change at the review comments?

File: crates/env/src/backend.rs (hunk 2 above)
Comment (kziemianek) [reply crates/env/src/backend.rs:298]: done

**Comment by kziemianek** [comment]
> Hey @kziemianek did you have a change at the review comments?



> Hey @kziemianek did you have a change at the review comments?

Yes, they all are resolved.

**Comment by goastler** [comment]
I have fixed the unhandled errors I was having with this PR, added tests, added docs and fixed the merge conflicts in my PR (#1840) that builds upon this one

File: crates/env/src/engine/on_chain/ext/wasm32.rs [hunk 4]
```diff
@@ -146,6 +146,15 @@ mod sys {
             output_ptr: Ptr32Mut<[u8]>,
         ) -> ReturnCode;
 
+        /// **WARNING**: this function is from the [unstable interface](https://github.com/paritytech/substrate/tree/master/frame/contracts#unstable-interfaces),
+        /// which is unsafe and normally is not available on production chains.
+        pub fn sr25519_verify(
```
Comment (xgreenx) [inline crates/env/src/engine/on_chain/ext/wasm32.rs:151]: @athei @cmichi What do you think about adding a new `unstable` feature to the `ink_env` crate?

File: crates/env/src/engine/off_chain/impls.rs (hunk 3 above)
Comment (xgreenx) [reply crates/env/src/engine/off_chain/impls.rs:348]: Could you create a `SIGNING_CTX` constant like in the substrate and add a comment with the link to the code on the substrate side(that substrate uses the same context)?

File: crates/e2e/macro/src/config.rs [hunk 5]
```diff
@@ -63,7 +63,7 @@ impl TryFrom<ast::AttributeArgs> for E2EConfig {
                     return Err(format_err_spanned!(
                         arg,
                         "expected a string literal for `additional_contracts` ink! E2E test configuration argument",
-                    ))
+                    ));
```
Comment (xgreenx) [inline crates/e2e/macro/src/config.rs:66]: Could you exclude this file from the PR, please?=)

File: crates/ink/ir/src/ir/item_mod.rs [hunk 6]
```diff
@@ -263,7 +263,7 @@ impl ItemMod {
                         .into_combine(format_err!(
                             overlap.span(),
                             "first ink! message with overlapping wildcard selector here",
-                        )))
+                        )));
```
Comment (xgreenx) [inline crates/ink/ir/src/ir/item_mod.rs:266]: Could you exclude this file from the PR, please?=)

File: crates/ink/ir/src/ir/trait_def/item/mod.rs [hunk 7]
```diff
@@ -384,7 +384,7 @@ impl InkItemTrait {
                 ).into_combine(format_err_spanned!(
                     duplicate_selector,
                     "first ink! trait constructor or message with same selector found here",
-                )))
+                )));
```
Comment (xgreenx) [inline crates/ink/ir/src/ir/trait_def/item/mod.rs:387]: Could you exclude this file from the PR, please?=)

File: crates/ink/src/env_access.rs [hunk 8]
```diff
@@ -880,6 +880,16 @@ where
             .map_err(|_| Error::EcdsaRecoveryFailed)
     }
 
+    pub fn sr25519_verify(
```
Comment (xgreenx) [inline crates/ink/src/env_access.rs:883]: Could you add a comment here as well, please?=)

**Comment by xgreenx** [comment]
Closed in favor of https://github.com/paritytech/ink/pull/1840

**Pull request closed without merging**